### DISCUSSION TITLE
How can i change the polymorphic identity of an object on the fly ?

### DISCUSSION THREAD

**Discussion by FatfoutiMontassar** [sqlalchemy/sqlalchemy#11036] [Usage Questions] [answered]
Hello,
i have this design:

```
class parent(SqlAlchemyBase):
    __tablename__ = "parent"
    __mapper_args__ = {"polymorphic_on": "type"}

    id = Column(table.UUID, primary_key=True)
    type = Column(
        EnumByValue(
            type, name="type"
        ),
        default=type.child1,
        server_default=type.child1.value,
        nullable=False,
    )


class child1(parent):
    __tablename__ = "child1"
    __mapper_args__ = {
        "polymorphic_identity": type.child1.value
    }

    id = Column(
        table.UUID,
        ForeignKey("parent.id"),
        primary_key=True,
    )


class child2(ItemConfigurationRecipient):
    __tablename__ = "child2"
    __mapper_args__ = {
        "polymorphic_identity": type.child2.value
    }

    id = Column(
        table.UUID,
        ForeignKey("parent.id"),
        primary_key=True,
    )
```
    
what i want to do is to change the type of a child while keeping the same ID

let's suppose i have object = (id:1, type:'child1')
when i do:

```
object.type = 'child2'
session.merge(object)
```

i get a StaleDataError:
`sqlalchemy.orm.exc.StaleDataError: UPDATE statement on table 'journey_item_configuration_recipient_stakeholder' expected to update 1 row(s); 0 were matched `

Any hints about what i'm doing wrong please ?
Thanks

**Comment by zzzeek** [comment, marked as answer]
the object's class is part of its identity key.    so you can't just change "type" like that, you would need to also change the object's class.    Then, since you are using joined table inheritance, SQLAlchemy has no logic to emit a DELETE from the "child1" table, you have to do that manually, and you would also need to emit an INSERT for the new child table manually.  SQLAlchemy doesnt support any of these operations directly.

overall it's much easier to delete the object, make a new one with the class you want, then add it again.

**Reply by FatfoutiMontassar** [reply]
Thanks for the answer 👍
